model/base_code_model.py

```diff
@@ -0,0 +1,158 @@
+from abc import ABCMeta, abstractmethod
+from math import ceil
+from typing import Tuple, Dict, List, Union
+
+import torch
+from pytorch_lightning.core.lightning import LightningModule
+from torch.optim import Adam, Optimizer, SGD
+from torch.optim.lr_scheduler import ExponentialLR, _LRScheduler
+from torch.utils.data import DataLoader
+
+from configs import ModelHyperparameters
+from dataset import Vocabulary, create_dataloader, PathContextBatch
+from utils.metrics import SubtokenStatistic, ClassificationStatistic
+
+StatisticType = Union[SubtokenStatistic, ClassificationStatistic]
+
+
+class BaseCodeModel(LightningModule, metaclass=ABCMeta):
+    def __init__(
+        self, config: ModelHyperparameters, vocab: Vocabulary, num_workers: int = 0,
+    ):
+        super().__init__()
+        self.save_hyperparameters()
+        self.config = config
+        self.vocab = vocab
+        self.num_workers = num_workers
+
+    def forward(
+        self,
+        samples: Dict[str, torch.Tensor],
+        paths_for_label: List[int],
+        output_length: int = None,
+        target_sequence: torch.Tensor = None,
+    ) -> torch.Tensor:
+        return self.decoder(self.encoder(samples), paths_for_label, output_length, target_sequence)
+
+    @abstractmethod
+    def _calculate_loss(self, logits: torch.Tensor, labels: torch.Tensor) -> torch.Tensor:
```

Comment:
As I said in slack, this method used only in `code2seq`. For `code2class` you can call `CrossEntropyLoss` in one line, you don't need to move it in different method.